From 565241067e7328fdee0d06acf5aee2c2311c9262 Mon Sep 17 00:00:00 2001
From: Seunghui Lee <sh043.lee@samsung.com>
Date: Thu, 17 Jul 2025 17:12:13 +0900
Subject: [PATCH] scsi: ufs: core: Use link recovery when h8 exit fails during
 runtime resume

[ Upstream commit 35dabf4503b94a697bababe94678a8bc989c3223 ]
[ Follow-up 62c015373e1c: only recover on hibern8 *exit* in wl_resume ]

If the h8 exit fails during runtime resume, skip EH while a PM op is
in progress and run link recovery only from __ufshcd_wl_resume().
Do not recover from ufshcd_uic_pwr_ctrl(): that also runs on hibern8
enter during suspend.

Fixes: 4db7a2360597 ("scsi: ufs: Fix concurrency of error handler and other error recovery paths")
Signed-off-by: Seunghui Lee <sh043.lee@samsung.com>
Signed-off-by: Peter Wang <peter.wang@mediatek.com>
---
diff --git a/drivers/ufs/core/ufshcd.c b/drivers/ufs/core/ufshcd.c
--- a/drivers/ufs/core/ufshcd.c
+++ b/drivers/ufs/core/ufshcd.c
@@ -4311,7 +4311,7 @@ static int ufshcd_uic_pwr_ctrl(struct ufs_hba *hba, struct uic_command *cmd)
 	hba->uic_async_done = NULL;
 	if (reenable_intr)
 		ufshcd_enable_intr(hba, UIC_COMMAND_COMPL);
-	if (ret) {
+	if (ret && !hba->pm_op_in_progress) {
 		ufshcd_set_link_broken(hba);
 		ufshcd_schedule_eh_work(hba);
 	}
@@ -9836,7 +9836,15 @@ static int __ufshcd_wl_resume(struct ufs_hba *hba, enum ufs_pm_op pm_op)
 		} else {
 			dev_err(hba->dev, "%s: hibern8 exit failed %d\n",
 					__func__, ret);
-			goto vendor_suspend;
+			/*
+			 * If the h8 exit fails during the runtime resume
+			 * process, it becomes stuck and cannot be recovered
+			 * through the error handler. To fix this, use link
+			 * recovery instead of the error handler.
+			 */
+			ret = ufshcd_link_recovery(hba);
+			if (ret)
+				goto vendor_suspend;
 		}
 	} else if (ufshcd_is_link_off(hba)) {
 		/*
